lib/prism/translation/parser/lexer.rb

```diff
@@ -321,17 +321,39 @@ def to_a
                   location = Range.new(source_buffer, offset_cache[next_location.start_offset], offset_cache[next_location.end_offset])
                   index += 2
                 end
-              elsif value.start_with?("<<")
+              elsif token.type == :HEREDOC_START
                 quote = value[2] == "-" || value[2] == "~" ? value[3] : value[2]
+                heredoc_type = value[2] == "-" || value[2] == "~" ? value[2] : ""
+                heredoc = HeredocData.new(
+                  identifier: value.match(/<<[-~]?["'`]?(?<heredoc_identifier>.*?)["'`]?\z/)[:heredoc_identifier],
+                  common_whitespace: 0,
+                )
+
                 if quote == "`"
                   type = :tXSTRING_BEG
                   value = "<<`"
                 else
+                  # The parser gem trims whitespace from squiggly heredocs. We must record
+                  # the most common whitespace to later remove.
+                  if heredoc_type == "~"
+                    heredoc.common_whitespace = calculate_heredoc_whitespace(index)
+                  end
+
                   value = "<<#{quote == "'" || quote == "\"" ? quote : "\""}"
                 end
+
+                heredoc_stack.push(heredoc)
               end
             when :tSTRING_CONTENT
-              unless (lines = token.value.lines).one?
+              if (lines = token.value.lines).one?
+                # Heredoc interpolation can have multiple STRING_CONTENT nodes on the same line.
+                is_first_token_on_line = lexed[index - 1] && token.location.start_line != lexed[index - 2][0].location&.start_line
+                # The parser gem only removes indentation when the heredoc is not nested
+                not_nested = heredoc_stack.size == 1
+                if is_first_token_on_line && not_nested && (heredoc = heredoc_stack[0]).common_whitespace > 0
```

Comment:
Any tip for getting this through sorbet? It complains that `heredoc` can be nil but that will never be the case because the previous condition checked that the first element is present.

When rewriting this into something that proves to sorbet that it can't be nil, it's telling me that the condition is always falsy.

Reply:
Got a workaround from https://github.com/sorbet/sorbet/issues/8401#issuecomment-2559830955 that seems to work